### ISSUE TITLE
Visualization gets messy in some cases

### ISSUE THREAD

**Issue by samuell** [spotify/luigi#1283]
We have a problem that with certain dependency graph topologies, they get very very very long (on the page) ... seemingly due to how text labels are shifted to be stacked diagonally upon each other.

In our case, this seems to create a "domino" effect, with longer and longer connections between up/downstream stages, to the extent that it gets unreadable. See example here (this is only the "top of the iceberg" :P ): 
![image](https://cloud.githubusercontent.com/assets/125003/10343710/7145c920-6d1f-11e5-8f84-44641a126c75.png)

**Comment by erikbern** [comment]
did you try the new visualizer?

**Comment by samuell** [comment]
@erikbern Ah, no ... that's news to me ...gotta check ... any pointer?

**Comment by Tarrasch** [comment]
@samuell it's in Luigi HEAD. Not on pypi yet.

Though you'll still have this issue in the svg visualiser

**Comment by samuell** [comment]
@Tarrasch Ok, thanks ... will have a look though!

**Comment by Tarrasch** [comment]
I don't know how to solve this though. @stephenpascoe and @daveFNbuck were talking about introducing the new filtering on the graphs too. It could solve this problem.

**Comment by daveFNbuck** [comment]
I have this problem as well, and the D3 graph doesn't help because it's way too slow on larger graphs. Filtering by status wouldn't really help in the pictured example, as the graph is all pending nodes anyway. My graphs tend to be dominated by done nodes and have a fairly pending subgraph. One thing that I think could help here is to add a depth limit to the graph. You won't be able to see everything at once, but it'll load much faster and you can always click on the nodes you're interested in to see more detail. It would be much more useful than a lot of my graphs in which 99% of the scrolling is just pages and pages of lines with no nodes.

**Comment by daveFNbuck** [comment]
Also, we should be able to swap in a better graph layout algorithm. There's no reason that the portion of the graph shown in this example needs to use so much vertical space.

**Comment by samuell** [comment]
Good description of the problem, @daveFNbuck !

But then, comparing to some other workflow tools, like [galaxy](http://galaxyproject.org), I realize luigi does visualization on the task level, while I think many other tools just do it on the process level (one box for each task *class*, not each *instance*) ...

**Comment by samuell** [comment]
... that is, one box per task class, with indicators telling how many instances of that task class are pending, running finished etc.

**Comment by samuell** [comment]
Or maybe, only on clicking one such, it is expanded to task instances. THat would be a sort of "defult grouping". Would be cool!

**Comment by daveFNbuck** [comment]
Someone on my team recently requested the same thing, but it would be tough for workflows where every task has the same class. Some of my most complicated graphs come from impala/hive queries that all share a common class but differ based on which tables they read from and write to.

**Comment by samuell** [comment]
True! Actually we have a planned feature in sciluigi where we will get the same problem as well ... having the same ShellTask class for all shell based tasks, which is close to all of them. Good point.

**Comment by daveFNbuck** [comment]
This could work if we indicated which parameters were suitable for ignoring when grouping nodes. For me, it would be fine to group together nodes with the same class and output table but different dates. If we marked the date parameter as ignore_for_grouping or something like that, it would allow the graph to be greatly simplified.

**Comment by daveFNbuck** [comment]
This might have a fairly simple solution. From graph.js, this is how the nodes are laid out vertically:

    /* Format nodes according to their depth and horizontal sort order.
       Algorithm: evenly distribute nodes along each depth level, offsetting each
       by the text line height to prevent overlapping text. The height of each
       depth level is therefore determined by the number of nodes at that depth. */

Looking at the task graph example, we see that this isn't really necessary. The first three nontrivial layers are spread out enough that we don't need to shift to avoid overlaps. In the next layer, we could get away with just two heights, alternating nodes like this:

```
node1   node3   node5
     node2   node4   node6
```

in the next layer down where the effect starts to get out of control, there's some secondary effect where each group is pushed down further. If we eliminate that effect and group them in tens or so, we could dramatically reduce the necessary vertical space. This will break down a bit with extremely wide levels, but we can break them into multiple rows or just stop including labels when this happens.

**Comment by daveFNbuck** [comment]
Alternatively, we can do a simple fixed-width layout for each level and just stack them, like

```
node1    node4   node7
node2    node5
node3    node6
```

that would probably be simpler.

**Comment by samuell** [comment]
Yeah, I think you're on to something here, @daveFNbuck , as there is indeed excess, unnecessary stacking happening there, which seems to be the main problem.

**Comment by samuell** [comment]
For grouping, there is already a parameter to decide whether a parameter should help make the task unique ... I think it is named "distinct" or something ...

**Comment by samuell** [comment]
Ok, I found now that I could go from this:
![image](https://cloud.githubusercontent.com/assets/125003/10348267/5cf37eb8-6d38-11e5-8a59-455986c86567.png)

... to this:
![image](https://cloud.githubusercontent.com/assets/125003/10348272/6124aae8-6d38-11e5-908c-c894ebe1b07a.png)

... with the following change in graph.js:

```diff
diff --git a/luigi/static/visualiser/js/graph.js b/luigi/static/visualiser/js/graph.js
index cfa0831..c477bbc 100644
--- a/luigi/static/visualiser/js/graph.js
+++ b/luigi/static/visualiser/js/graph.js
@@ -83,11 +83,11 @@ Graph = (function() {
     function layoutNodes(nodes, rowSizes) {
         function rowStartPosition(depth) {
             if (depth === 0) return 20;
-            return rowStartPosition(depth-1)+Math.max(rowSizes[depth-1]*10+10,100);
+            return rowStartPosition(depth-1)+Math.max(rowSizes[depth-1],100);
         }
         $.each(nodes, function(i, node) {
             node.x = ((node.xOrder+1)/(rowSizes[node.depth]+1))*(graphWidth-200)+100;
-            node.y = rowStartPosition(node.depth) + (node.xOrder*10);
+            node.y = rowStartPosition(node.depth) + ((node.xOrder % 100) * 10);
         });
     }
```

**Comment by samuell** [comment]
... so, the modulo on the latter line avoids the "infinite stacking", while the change on the former line avoids making row sizes "explode" too much with deeper depths.

**Comment by samuell** [comment]
Ok, for a more dense graph, even more changes were needed:

```diff
diff --git a/luigi/static/visualiser/js/graph.js b/luigi/static/visualiser/js/graph.js
index cfa0831..716b04c 100644
--- a/luigi/static/visualiser/js/graph.js
+++ b/luigi/static/visualiser/js/graph.js
@@ -83,11 +83,11 @@ Graph = (function() {
     function layoutNodes(nodes, rowSizes) {
         function rowStartPosition(depth) {
             if (depth === 0) return 20;
-            return rowStartPosition(depth-1)+Math.max(rowSizes[depth-1]*10+10,100);
+            return rowStartPosition(depth-1)+100;
         }
         $.each(nodes, function(i, node) {
             node.x = ((node.xOrder+1)/(rowSizes[node.depth]+1))*(graphWidth-200)+100;
-            node.y = rowStartPosition(node.depth) + (node.xOrder*10);
+            node.y = rowStartPosition(node.depth) + (node.xOrder % 100)*10;
         });
     }
```

... but then it gets a little messy in another way (... though if you ask me, it is still better than those infinite vertical lines):

![selection_368](https://cloud.githubusercontent.com/assets/125003/10349131/3f03fa0e-6d3d-11e5-84de-46e2268ac8cd.png)

**Comment by daveFNbuck** [comment]
@samuell try your graph with #1286, it should be a bit better

**Comment by daveFNbuck** [comment]
@samuell You made your rowStartPosition way too small, you want

    return rowStartPosition(depth-1)+Math.max(Math.min(rowSizes[depth-1], 100)*10+10,100);

to get enough room between rows by limit it to the 100 your modulus gives you.

**Comment by samuell** [comment]
@daveFNbuck Cool, with the two patches, my very entangled graph becomes:

![selection_377](https://cloud.githubusercontent.com/assets/125003/10354322/49f51750-6d5f-11e5-86a9-0a2c1553c130.png)

Obviously my close-to-non-existent javascript skills failed me here, but this is awesome, and a huge improvement! :)

**Comment by daveFNbuck** [comment]
I think you were close, but small errors get magnified on a big graph like that.

**Comment by daveFNbuck** [comment]
The guy who was bugging me about this on my team is happy now too, so my fixes seem to work well on real graphs.

**Comment by samuell** [comment]
:+1: :)

**Comment by LongbinChen** [comment]
I am "the guy". Thanks Dave.

**Comment by Tarrasch** [comment]
Merged #1286 and #1287. So closing this one. Huge thanks to @daveFNbuck! Also thanks to Samuel and "the guy" for reporting this. ;)
